app.py

```diff
@@ -264,245 +295,133 @@ def summarize_marriage_bonus(marriage_bonus):
-        "Benefits": "Benefits Change %",
-        "Taxes": "Tax Change %",
-        "Credits": "Credit Change %"
+    for i, credits in enumerate(refundable_credits):
+        if credits_married[i] != 0 or credits_separate[i] != 0:
+            credits_table["Program"].append(format_program_name(credits))
+            credits_table["Not Married"].append(formatted_credits_separate[i])
+            credits_table["Married"].append(formatted_credits_married[i])
+            credits_table["Delta"].append(formatted_credits_delta[i])
+            credits_table["Delta Percentage"].append(
+                formatted_credits_delta_percent[i]
+            )
+
+    refundable_credits_df = pd.DataFrame(credits_table)
+
+    taxes_table = {
+        "Program": [],
+        "Not Married": [],
+        "Married": [],
+        "Delta": [],
+        "Delta Percentage": [],
     }
-    abs_max = max(abs(min(map(min, data))), abs(max(map(max, data))))
-    z_min = -abs_max
-    z_max = abs_max
-    color_scale = [
-            (0, '#616161'), 
-            (0.5, '#FFFFFF'),  
-            (1, '#2C6496')  
-            ]
-    
-    
-    
-    
-    # Display the chart once data calculation is complete
-    fig = px.imshow(data,
-                    labels=dict(x="Head Employment Income", y="Spouse Employment Income", color=label_legend[heatmap_tax_unit]),
-                    x=x_values,
-                    y=y_values,
-                    zmin=z_min,
-                    zmax=z_max,
-                    color_continuous_scale=color_scale,
-                    origin='lower'
-                )
-    
-    # Add custom data (percentage values) to the figure
-    fig.update_traces(customdata=data)
-    
-    fig.update_xaxes(side="bottom")
-    fig.update_layout(
-        xaxis=dict(
-            tickmode='array',
-            tickvals=[0, 10000, 20000, 30000, 40000, 50000, 60000, 70000, 80000],
-            ticktext=["{}k".format(int(val/1000)) for val in [0, 10000,20000, 30000,40000,50000, 60000, 70000, 80000]],
-            showgrid=True,
-            zeroline=False,
-            title=dict(text='Head Employment Income', standoff=15),
-        ),
-        yaxis=dict(
-            tickmode='array',
-            tickvals=[0, 10000, 20000, 30000, 40000, 50000, 60000, 70000, 80000],
-            ticktext=["{}k".format(int(val/1000)) for val in [0, 10000, 20000, 30000, 40000, 50000, 60000, 70000, 80000]],
-            showgrid=True,
-            zeroline=False,
-            title=dict(text='Spouse Employment Income', standoff=15),
-            scaleanchor="x",
-            scaleratio=1,
-        )
-    )
-   
-    fig.update_layout(height=600, width=800)
-    
-    # Customize hover template to display x and y values along with the percentage change
-    fig.update_traces(hovertemplate='Head Employment Income: %{x}<br>Spouse Employment Income: %{y}<br>Change: %{customdata:.2f}%')
 
+    taxes_before_refundable_credits = list(programs[0][-3].keys())
```

Comment:
you can avoid code duplication by creating a single dataframe with all the variables you want across the three situations, and adding a column for the category (Total, Taxes before refundable credits, Refundable credits, Benefits) then when displaying the breakdown filtering the table on that category. Then you can calculate absolute and relative differences in one dataframe command.